src/factor_instances_provider/provider/factor_instances_cache.rs

```diff
@@ -0,0 +1,525 @@
+/// it in Sargon.
+///
+/// We use `IndexMap` instead of `HashMap` for future proofing when we serialize,
+/// deserialize this cache, we want the JSON values to have stable ordering. Note
+/// that the only truly **important** ordering is that of `FactorInstances` values,
+/// which are ordered since it is a newtype around `IndexSet<HierarchicalDeterministicFactorInstance>`.
+///
+///
+/// The Serde impl of `IndexAgnosticPath` could be:
+/// `"<Network>/<EntityKind>/<KeyKind>/<KeySpace>"` as a string, e.g:
+/// `"1/A/TX/U"`, where `U` is "Unsecurified" KeySpace.
+/// Or if we don't wanna use such a "custom" one we can use `525`/`616`
+/// discriminator for EntityKind and `1460`/`1678` for KeyKind:
+/// "1/525/1460/U".
+#[derive(Debug, Default, Clone)]
+pub struct FactorInstancesCache {
+    /// PER FactorSource PER IndexAgnosticPath FactorInstances (matching that IndexAgnosticPath)
+    map: IndexMap<FactorSourceIDFromHash, IndexMap<IndexAgnosticPath, FactorInstances>>,
+}
+
+impl FactorInstancesCache {
+    /// Inserts `instances` under `factor_source_id` by splitting them and grouping
+    /// them by their `IndexAgnosticPath`.
+    ///
+    /// Returns `Err` if any of the instances is in fact does NOT have `factor_source_id`,
+    /// as its factor source id.
+    ///
+    /// Returns `bool` indicating if an index was skipped resulting in non-contiguousness, which
+    /// we do not use for now. Might be something we enforce or not for certain operations
+    /// in the future.
+    pub fn insert_for_factor(
+        &mut self,
+        factor_source_id: &FactorSourceIDFromHash,
+        instances: &FactorInstances,
+    ) -> Result<bool> {
+        let mut skipped_an_index_resulting_in_non_contiguousness = false;
+
+        let instances_by_agnostic_path = InstancesByAgnosticPath::from(instances.clone());
+        instances_by_agnostic_path.validate_from_source(factor_source_id)?;
+        if let Some(existing_for_factor) = self.map.get_mut(factor_source_id) {
+            for (agnostic_path, instances) in instances_by_agnostic_path {
+                let instances = instances.factor_instances();
+
+                if let Some(existing_for_path) = existing_for_factor.get_mut(&agnostic_path) {
+                    if let Some(fi) = instances
+                        .intersection(&existing_for_path.factor_instances())
+                        .next()
+                    {
+                        return Err(CommonError::CacheAlreadyContainsFactorInstance {
+                            derivation_path: fi.derivation_path(),
+                        });
+                    }
+
+                    if let Some(last) = existing_for_path.factor_instances().last() {
+                        if instances.first().unwrap().derivation_entity_base_index()
+                            != last.derivation_entity_base_index() + 1
+                        {
+                            warn!(
```

Comment:
We typically want all indices to be contiguous - as in no "gaps" - but in case of legacy Profile that will not be true, so we can typically not enforce it. Perhaps this `warn!` should be changed to `info!`....